### ISSUE TITLE
[BUG]: If title of article changes, the article adds as new one

### ISSUE THREAD

**Issue by sakkamade** [martinrotter/rssguard#451]
#### Brief description of the issue.

When certain article has already been fetched and this article's title changes, a separate article with changed title appends to the list as new entry.

#### How to reproduce the bug?
1. E.g. copy the link from `Atom topic feed` at the bottom of [bbs.archlinux.org](https://bbs.archlinux.org/viewtopic.php?id=268158) page (or any similar)
2. Fetch it through RSS Guard
3. Wait until it gets `[SOLVED]`

#### What was the expected result?

The title of already existing article to be amended.

#### What actually happened?

Identical (or slightly distinct) article with changed title is being added to the list.

#### Other information

Akregator is capable of it.

I guess the criteria of such filtering (whether to add new entry or amend already present one) is being the link to the article, which stays invariant. 

* OS: Arch Linux, KDE
* RSS Guard
Version: 4.0.0 (built on Linux/x86_64)
Revision: 02993087
Build date: 16/07/2021 06:16
Qt: 5.15.2 (compiled against 5.15.2)

**Comment by martinrotter** [comment]
@sakkamade You are right. There was lengthy discussions about this some many years ago and it was problematic.

[Here](https://github.com/martinrotter/rssguard/blob/master/src/librssguard/database/databasequeries.cpp#L1048) is the code.

Firstly, some feed types have special `ID` which uniquely identifies the article and stays the same if (for example) title changes but still describes same event (for example):
https://datatracker.ietf.org/doc/html/rfc4287#section-4.2.6
https://validator.w3.org/feed/docs/rss2.html#hrelementsOfLtitemgt
https://jsonfeed.org/version/1.1

RSS Guard idenitifes the article as "the same" if it has same URL and (at the same time) TITLE and AUTHOR. IDs provided by the feed are not taken into account (for synchronized services like TT-RSS this is different). I remember that this state was de-facto result of "that" discussion we had with some users in RSS Guard prehistoric times and maybe it is time to re-do it.

Thing also is that many feeds do not provide IDs for their articles.

So basically we can choose from:

1. If article provides ID, then use that ID as unique idenifier.
2. Do not take TITLE of article into account when comparing them for uniqueness.

And/or I could enhance "message filtering" mechanism to be able to:
* find existing article (and potentially overwrite it directly) in RSS Guard's DB with input parameters (URL OR AUTHOR, etc.) and return its data (useful in cooperation with 1. - you could "generate" IDs for articles which do not have them, for example use their URL as ID, or you could change title/author/whatever directly in the DB if article is already there, ...,

There are many options, too many.

**Comment by sakkamade** [comment]
> some feed types have special ID which uniquely identifies the article

Source on it, or sample of how it looks? For I am not quite sure I have ever seen it.

---

> So basically we can choose from:
> 1. If article provides ID, then use that ID as unique idenifier.
> 1. Do not take TITLE of article into account when comparing them for uniqueness.

For myself, I would prefer `#1` and 

> ... enhance "message filtering" mechanism ...

However, at the same time, I cannot come upon an example of situation where user will actually benefit from it.
So the second would be for best, I guess?

**Comment by sakkamade** [comment]
> > some feed types have special ID which uniquely identifies the article
> Source on it, or sample of how it looks? For I am not quite sure I have ever seen it.

Never mind. Most (every single?) of the feeds I have seen have it. In every feed I have see, it contains the link on the article.

Now I see what you mean.
And now I am of opinion that `#1` is the best option.

**Comment by martinrotter** [comment]
@sakkamade 

![image](https://user-images.githubusercontent.com/1255302/126889001-a8b6a433-e006-4d56-ab08-a1a776b81291.png)

But note that "guid" does not have to be URL. It can be just about any string, but is URL in many cases indeed.

Yes, other feed readers usually utilise "guid", I will try to make proposed changes.

**Comment by martinrotter** [comment]
ok, pushed the change, feel free to test with latest devbuild

**Comment by martinrotter** [comment]
@sakkamade The change proved to be more difficult than I anticipated because "ID" of message is used on many places throughout RSS Guard (particularly in cooperation with synchronized services) so I had to specially take care of some corner cases, hopefully it will work out nicely.

Let's give this change some time, report any associated bugs here.

**Comment by sakkamade** [comment]
Well, I don't know...

Something appears to have broken, but the problem is awfully specific, so I will provide a concrete example:

- Feed link: 
   https://github.com/dvorka.atom
- Problematic article: 
   "dvorka created a branch dev/2.4.0 in dvorka/hstr"

### Issue `#1`
 Vague steps to reproduce:
   1. Fetch the feed
   2. Mark all articles as read
   3. Wait some time
   3. Fetch the feed again
   4. See aforementioned problematic article becoming **un**read

### Issue `#2`
 Steps to reproduce (I am not quite sure what is happening here at all):
   1. Sort the articles by Date
   2. Fetch the feed couple of times
   3. See how articles jumping about (the dates are constant, but sorting is changing all the same)

---

RSS Guard
Version: 4.0.0 (built on Linux/x86_64)
Revision: 68be51b9
Build date: 27/07/2021 03:02
Qt: 5.15.2 (compiled against 5.15.2)

**Comment by martinrotter** [comment]
ok will check,

**Comment by martinrotter** [comment]
@sakkamade `Wait some time` How long?

**Comment by martinrotter** [comment]
Issue 2 - really, it seems that the problem is somehow triggered if `CTRL` is part of shortcut. If I remap the action for example to `F6` key, it works okay.

**Comment by sakkamade** [comment]
>  If I remap the action for example to `F6` key, it works okay. Will investigate.

Ah, and here I was thinking how I have never noticed it before... 
I am using `F5`, on daily basis :heart:

**Comment by martinrotter** [comment]
OK, Issue 1 - in your particular "dvorka" feed, the Github seem to change CONTENTS of message often.
![image](https://user-images.githubusercontent.com/1255302/127103400-78d38311-5b30-49ce-83f0-b69582c41e47.png)

Maybe, RSS Guard should update and mark as unread only if title changes? Not contents? Maybe message should NOT be marked as unread again if such an update happens?

**Comment by sakkamade** [comment]
> So RSS Guard did it right. Message CONTENTS was changed, message was UPDATED and marked as UNREAD again.

I see!

> Maybe, RSS Guard should update and mark as unread only if title changes? Not contents? Maybe message should NOT be marked as unread again if such an update happens?

By no means! Current behaviour should be left as is, in this case.
Thank you.

I will open issue `#2` as separate then.

**Comment by martinrotter** [comment]
> I will open issue `#2` as separate then.

Yes, do that. I will make some more testing for potential issue 1.

**Comment by martinrotter** [comment]
Wow, that's crazy. I fetched THE feed once more and now Github returns
![image](https://user-images.githubusercontent.com/1255302/127104100-0650387e-752b-4fd9-9a2e-664daebb60a8.png)

The bottom part of contents got changed in feed again to "Updated Jul 24". WTF.

**Comment by sakkamade** [comment]
> Wow, that's crazy. I fetched THE feed once more and now Github returns
> The bottom part of contents got changed in feed again to "Updated Jul 24"

Yes, that is the thing, it happens every 2-3, sometimes even less than 1, minutes. 
(Oops, forgot to mention this...)

**Comment by martinrotter** [comment]
Yes, they keep rotating 24-25. WTF.

**Comment by martinrotter** [comment]
reported https://github.community/t/atom-feed-returns-inconsistent-data/192296
